src/trace_ng/log.zig

```diff
@@ -0,0 +1,273 @@
+        exit_sig: AtomicBool,
+        allocator: Allocator,
+        free_fba: RecycleFBA,
+        fba_bytes: u64,
+        channel: *Channel(logfmt.LogMsg),
+        handle: ?std.Thread,
+
+        pub fn init(config: Config) Self {
+            return .{
+                .allocator = config.allocator,
+                .free_fba = RecycleFBA.init(config.allocator, config.fba_bytes) catch @panic("could not create RecycleFBA"),
+                .fba_bytes = config.fba_bytes,
+                .level = config.level,
+                .exit_sig = AtomicBool.init(false),
+                .channel = Channel(logfmt.LogMsg).init(config.allocator, INITIAL_LOG_CHANNEL_SIZE),
+                .handle = null,
+            };
+        }
+
+        fn unscoped(self: Self) UnscopedLogger {
+            return .{
+                .level = self.level,
+                .exit_sig = self.exit_sig,
+                .allocator = self.allocator,
+                .free_fba = self.free_fba,
+                .fba_bytes = self.fba_bytes,
+                .channel = self.channel,
+                .handle = self.handle,
+            };
+        }
+
+        fn withScope(self: Self, comptime new_scope: anytype) StandardLogger(new_scope) {
+            return .{
+                .level = self.level,
+                .exit_sig = self.exit_sig,
+                .allocator = self.allocator,
+                .free_fba = self.free_fba,
+                .fba_bytes = self.fba_bytes,
+                .channel = self.channel,
+                .handle = self.handle,
+            };
+        }
+
+        pub fn spawn(self: *Self) void {
+            self.handle = std.Thread.spawn(.{}, Self.run, .{self}) catch @panic("could not spawn Logger");
+        }
+
+        pub fn deinit(self: *Self) void {
+            self.channel.close();
+            if (self.handle) |handle| {
+                self.exit_sig.store(true, .seq_cst);
+                handle.join();
+            }
+            self.channel.deinit();
+            self.free_fba.deinit();
+        }
+
+        pub fn run(self: *Self) void {
+            while (!self.exit_sig.load(.seq_cst)) {
+                std.time.sleep(std.time.ns_per_ms * 5);
+                const messages = self.channel.drain() orelse {
+                    // channel is closed
+                    return;
+                };
+                defer self.channel.allocator.free(messages);
+
+                for (messages) |message| {
+                    logfmt.writeLog(message) catch @panic("logging failed");
+                }
+            }
+        }
+
+        fn createLogMessage(
+            _: Self,
+            free_fba: *RecycleFBA,
+            total_len: u64,
+            level: Level,
+            maybe_scope: ?[]const u8,
+            maybe_msg: ?[]const u8,
+            maybe_fields: anytype,
+            comptime maybe_fmt: ?[]const u8,
+            args: anytype,
+        ) logfmt.LogMsg {
+            // obtain a memory to write to
```

Comment:
This should realloc the buffer to a smaller size.

When you call free in the other thread, it's going to pass in a len that is smaller than the actual size of the allocation. For the RecycleFBA it does not actually make a difference. But with other allocators it may cause a problem. The GPA will complain about this.

```zig
test "invalid free" {
    const x = try std.testing.allocator.alloc(u8, 100);
    std.testing.allocator.free(x[0..50]);
}
```